src/core/proxy/src/listener.rs

```diff
@@ -3,181 +3,295 @@
+
+        let waker = Arc::new(Waker::from(
+            ::net::Waker::new(poll.registry(), WAKER_TOKEN).unwrap(),
+        ));
+
         let nevent = config.nevent();
         let timeout = Duration::from_millis(config.timeout() as u64);
 
-        let mut poll = Poll::new()?;
-        poll.bind(addr, tls_config)?;
+        let sessions = Slab::new();
 
         Ok(Self {
-            addr,
+            listener,
             nevent,
             poll,
+            sessions,
             timeout,
+            waker,
         })
     }
 
     pub fn waker(&self) -> Arc<Waker> {
-        self.poll.waker()
+        self.waker.clone()
     }
 
-    pub fn build(self, connection_queues: Queues<Session, ()>) -> Listener {
+    pub fn build(
+        self,
+        signal_queue: Queues<(), Signal>,
+        session_queue: Queues<Session, Session>,
+    ) -> Listener {
         Listener {
-            addr: self.addr,
-            connection_queues,
+            listener: self.listener,
             nevent: self.nevent,
             poll: self.poll,
+            sessions: self.sessions,
+            session_queue,
+            signal_queue,
             timeout: self.timeout,
+            waker: self.waker,
         }
     }
 }
 
 pub struct Listener {
-    addr: SocketAddr,
-    connection_queues: Queues<Session, ()>,
+    /// The actual network listener server
+    listener: ::net::Listener,
+    /// The maximum number of events to process per call to poll
     nevent: usize,
+    /// The actual poll instantance
     poll: Poll,
+    /// Sessions which have been opened, but are not fully established
+    sessions: Slab<Session>,
+    /// Queues for sending established sessions to the worker thread(s) and to
+    /// receive sessions which should be closed
+    session_queue: Queues<Session, Session>,
+    /// Queue for receieving signals from the admin thread
+    signal_queue: Queues<(), Signal>,
+    /// The timeout for each call to poll
     timeout: Duration,
+    /// The waker handle for this thread
+    waker: Arc<Waker>,
 }
 
 impl Listener {
-    /// Handle an event on an existing session
-    fn handle_session_event(&mut self, event: &Event) {
+    /// Call accept one time
+    fn accept(&mut self) {
+        if let Ok(mut session) = self.listener.accept().map(Session::from) {
+            if session.is_handshaking() {
+                let s = self.sessions.vacant_entry();
+                let interest = session.interest();
+                if session
+                    .register(self.poll.registry(), Token(s.key()), interest)
+                    .is_ok()
+                {
+                    s.insert(session);
+                } else {
+                    // failed to register
+                }
+            } else if self.session_queue.try_send_any(session).is_err() {
+                // we probably want to wake and retry here
+            }
+
+            // reregister is needed here so we will call accept if there is a backlog
```

Comment:
Everything is edge triggered in MIO. No option for level triggered. I wouldn't object to accepting up to a small number of connections here. The reason we do just one at a time is so accept does not delay SSL negotiation for sessions that are mid-handshake.

Reply:
Added loop within the accept call so that - accept is called up to 8 times or until an error occurs (eg: would block). If accept results in a session all 8 times, then we reregister the listener.